nemo_aligner/utils/distributed.py

```diff
@@ -17,29 +17,46 @@
 import time
 import warnings
 from collections import defaultdict
+from contextlib import contextmanager
 from dataclasses import dataclass
 from datetime import timedelta
 from typing import Dict, Optional, Union
 
 import torch
-from megatron.core import parallel_state, tensor_parallel
+from megatron.core import tensor_parallel
 
-from nemo.collections.nlp.modules.common.text_generation_utils import get_model_parallel_src_rank
 from nemo.utils.timers import NamedTimer
+from nemo_aligner.utils import parallel_state
+from nemo_aligner.utils.parallel_state import get_model_parallel_group, get_model_parallel_src_rank
 from nemo_aligner.utils.ppo_utils import calculate_entropy
 
 
-def gather_tensor(tensor, dst, group, dtype=torch.float32):
-    """Gather any 2d tensor to the dst rank from every other rank in the given group.
-    All the ranks that send or receive data must call this function."""
-    tensor = tensor.to(device=torch.cuda.current_device(), dtype=dtype)
-    if torch.distributed.get_rank() == dst:
-        gather_list = [torch.empty_like(tensor) for _ in range(torch.distributed.get_world_size(group))]
-    else:
-        gather_list = None
+def rebalance_nd_tensor(tensor, group):
+    """takes tensors with variable leading sizes(at dim=0) and then stack them into a single tensor. NOTE: assumes dim != 0 to have all the same shape
+    """
+    num_samples = torch.as_tensor(tensor.size(0), dtype=torch.int64, device=torch.cuda.current_device())
+    batch_num_per_rank = torch.zeros(
+        torch.distributed.get_world_size(group), dtype=torch.int64, device=torch.cuda.current_device()
+    )
+    torch.distributed.all_gather_into_tensor(batch_num_per_rank, num_samples, group=group)
+
+    B = batch_num_per_rank.sum()
+    other_dims = tensor.shape[1:]
+
+    indices = batch_num_per_rank.cumsum(dim=0)
+    output_tensor = torch.zeros(B, *other_dims, dtype=tensor.dtype, device=torch.cuda.current_device())
+
+    # tensor_split is a view we can copy into
+    output_tensor.tensor_split(indices.cpu())[torch.distributed.get_rank(group=group)].copy_(tensor)
```

Comment:
I suspect you can probably use `indices[0:-1]` here?

Reply:
hmm sorry i don't see it. do you mean replace the `tensor_split` function? or replace the `copy_`?

Reply:
I mean replacing `indices` with `indices[0:-1]`. This is because `tensor_split()` only needs the *start* indices of each sub-tensor, so passing `indices` in full also returns an empty final sub-tensor that we end up not using. Here's an example (compare the last outputs of the last two statements):
![image](https://github.com/NVIDIA/NeMo-Aligner/assets/507137/a3cc7dbe-b865-4bf7-8f84-614517131367)

Reply:
ah i see, and this `cumsum` will always end at the length the tensor so this indexing works.

3ecb522ccb320ce85fd70cd690c062aab1df61c6 fixes it